ui/pods/poddialogs/stats.go

```diff
@@ -0,0 +1,482 @@
+	return d
+}
+
+// SetPodsOptions sets pod drop down options
+func (d *PodStatsDialog) SetPodsOptions(options []PodStatsDropDownOptions) {
+	maxWidth := 0
+	d.podDropDownOptions = options
+	if len(options) == 0 {
+		return
+	}
+	podOptions := []string{"all"}
+	for i := 0; i < len(options); i++ {
+		item := options[i].ID
+		if options[i].Name != "" {
+			item = fmt.Sprintf("%s (%s)", item, options[i].Name)
+		}
+		if len(item) > maxWidth {
+			maxWidth = len(item)
+		}
+		podOptions = append(podOptions, item)
+	}
+	maxWidth = maxWidth + 11
+	d.controlLayout.ResizeItem(d.podDropDown, maxWidth, 0)
+	d.podDropDown.SetOptions(podOptions, d.setStatsQueryPodIDs)
+	d.podDropDown.SetCurrentOption(0)
+}
+
+func (d *PodStatsDialog) query() {
+	opts := d.getStatsQueryOptions()
+	podStats, err := ppods.Stats(opts)
+	if err != nil {
+		log.Error().Msgf("pod stats dialog: query error: %v", err)
+		return
+	}
+	d.updateData(podStats)
+}
+
+func (d *PodStatsDialog) startStatsQueryLoop() {
+	log.Debug().Msgf("pod stats dialog: starting pod stats query loop")
+	go func() {
+		tick := time.NewTicker(d.queryRefreshInterval)
+		// initial query
+		d.query()
+		for {
+			select {
+			case <-tick.C:
+				d.query()
+			case <-d.doneChan:
+				log.Debug().Msgf("pod stats dialog: stats query loop stopped")
+				return
+			}
+		}
+	}()
+}
+
+func (d *PodStatsDialog) getStatsQueryOptions() *ppods.StatsOptions {
+	d.mu.Lock()
+	opts := d.statQueryOpts
+	d.mu.Unlock()
+	return opts
+}
+
+func (d *PodStatsDialog) setStatsQueryPodIDs(name string, index int) {
+	if index == -1 {
+		return
+	}
+	d.mu.Lock()
+	if index == 0 {
+		d.statQueryOpts.IDs = d.getAllPodIDs()
+	} else {
+		d.statQueryOpts.IDs = []string{
+			d.podDropDownOptions[index-1].ID,
+		}
+	}
+	defer d.mu.Unlock()
+	go d.query()
+}
+
+func (d *PodStatsDialog) setStatsQuerySortBy(name string, index int) {
+	if index == -1 {
+		return
+	}
+	d.mu.Lock()
+	d.statQueryOpts.SortBy = index
+	defer d.mu.Unlock()
+	go d.query()
+}
+
+func (d *PodStatsDialog) getAllPodIDs() []string {
+	var ids []string
+	for _, item := range d.podDropDownOptions {
+		ids = append(ids, item.ID)
+	}
+	return ids
+}
+
+func (d *PodStatsDialog) initTableUI() {
+	var tableHeaders = []string{"POD ID", "CID", "NAME", "CPU %", "MEM USAGE / LIMIT", "MEM %", "NET IO", "BLOCK IO", "PIDS"}
+	headerBgColor := utils.Styles.PodStatsDialog.BgColor
+	headerFgColor := utils.Styles.PodStatsDialog.FgColor
+	d.table.Clear()
+	for index, header := range tableHeaders {
+		headerItem := fmt.Sprintf("[::b]%s[::-]", header)
+		d.table.SetCell(0, index,
+			tview.NewTableCell(headerItem).
+				SetExpansion(1).
+				SetAlign(tview.AlignLeft).
+				SetBackgroundColor(headerBgColor).
+				SetTextColor(headerFgColor).
+				SetSelectable(false))
+	}
+	d.table.SetFixed(1, 1)
+	d.table.SetSelectable(true, false)
+}
+
+func (d *PodStatsDialog) updateData(statReport []ppods.StatReporter) {
+	d.mu.Lock()
+	d.statsResult = statReport
+	d.mu.Unlock()
```

Comment:
Why isn't the defer right after the Lock? Of if all you want to lock is the section between the section above, why use defer?

Reply:
Here need section Lock, so I didn't use defer.